### PR TITLE
enable CrashLog::dumpStack outside Windows

### PR DESCRIPTION
**Pull request by Nindaleth**
This seems like the cleanest diff right now.

### CODE REVIEW

**Comment by Try** [comment]
Overall change look good to me, but can you please clarify a use-case?

**Comment by Nindaleth** [comment]
Well, it could at least save a file with info which signal killed the app and the little bit of system information.

I agree it would be better to extend edd-dbg to work on Linux, so that the `#ifdef` could be dropped. Does it compile and work on OSX/macOS, since edd-dbg advertises support for it?

**Comment by Nindaleth** [comment]
Edd Dawson said on his website that porting edd-db from OSX to Linux should be fairly easy, so I've tried it. I copied the osx dir into linux dir and fixed all compilation warnings and linker warnings. It shouldn't have been able to link on OSX because the library was never provided, not sure about the missing `-ldl`.

[WIP] because it doesn't really dump the stack yet (see below) and I don't know yet what's missing:

```
---crashlog(SIGSEGV)---
GPU: AMD RADV/ACO POLARIS10 (LLVM 10.0.1)
[no call frames available]

Segmentation fault (core dumped)
```

The important changes are into [`debugger_attached`](https://github.com/Try/OpenGothic/pull/94/files#diff-c97c1358aaff342e83471a2c33dc693981ca59307ecbcfcf378f6147b06aefe8R22) and [`frame_ptr_check` constructor](https://github.com/Try/OpenGothic/pull/94/files#diff-de9771f08675223b5b10a426bd09d182412b10ae040bbc988eb9c88fd972490dR49).

**Comment by Try** [comment]
> OSX to Linux should be fairly easy, so I've tried it. 

That's one way to do it, as alternative I want point out to linux 'execinfo.h': 
https://www.tutorialspoint.com/how-to-automatically-generate-a-stacktrace-when-a-gcc-cplusplus-program-crashes

**Comment by Nindaleth** [comment]
Thanks for the hint, the execinfo.h is promising. I've got some basic results already, but it's not demangled yet.

**Comment by Nindaleth** [comment]
Pushed an acceptable initial version. It's much better than nothing, but there are two weaknesses:
* hypothetical for now: the demangling using `__cxa_demangle` is non-standard, but GCC and LLVM support it (not much I can do about that right now, might need some ifdefs in case other niche compilers don't support). I didn't find explicit confirmation or denial, but I've seen some hints that ICC supports this too.
* actual weakness: source filename and line number not present in the output

Here's an example output (forced NULL pointer dereference in inventory window code):
```
---crashlog(SIGSEGV)---
GPU: AMD RADV POLARIS10 (ACO)
#1: InventoryMenu::keyUpEvent(Tempest::KeyEvent&) - /home/black_fox/src/OpenGothic/build/opengothic/Gothic2Notr(_ZN13InventoryMenu10keyUpEventERN7Tempest8KeyEventE+0x39) [0xa3b145]
#2: MainWindow::keyUpEvent(Tempest::KeyEvent&) - /home/black_fox/src/OpenGothic/build/opengothic/Gothic2Notr(_ZN10MainWindow10keyUpEventERN7Tempest8KeyEventE+0x1b2) [0x9e3ce0]
#3: Tempest::EventDispatcher::dispatchKeyUp(Tempest::Widget&, Tempest::KeyEvent&, unsigned int) - /home/black_fox/src/OpenGothic/build/opengothic/libMoltenTempest.so(_ZN7Tempest15EventDispatcher13dispatchKeyUpERNS_6WidgetERNS_8KeyEventEj+0x125) [0x7fd462b8bb93]
#4: Tempest::SystemApi::dispatchKeyUp(Tempest::Window&, Tempest::KeyEvent&, unsigned int) - /home/black_fox/src/OpenGothic/build/opengothic/libMoltenTempest.so(_ZN7Tempest9SystemApi13dispatchKeyUpERNS_6WindowERNS_8KeyEventEj+0x2d) [0x7fd462b9115d]
#5: Tempest::X11Api::implProcessEvents(Tempest::SystemApi::AppCallBack&) - /home/black_fox/src/OpenGothic/build/opengothic/libMoltenTempest.so(_ZN7Tempest6X11Api17implProcessEventsERNS_9SystemApi11AppCallBackE+0x4e8) [0x7fd462b869b2]
#6: Tempest::X11Api::implExec(Tempest::SystemApi::AppCallBack&) - /home/black_fox/src/OpenGothic/build/opengothic/libMoltenTempest.so(_ZN7Tempest6X11Api8implExecERNS_9SystemApi11AppCallBackE+0x46) [0x7fd462b864c0]
#7: Tempest::SystemApi::exec(Tempest::SystemApi::AppCallBack&) - /home/black_fox/src/OpenGothic/build/opengothic/libMoltenTempest.so(_ZN7Tempest9SystemApi4execERNS0_11AppCallBackE+0x27) [0x7fd462b912b7]
#8: Tempest::Application::exec() - /home/black_fox/src/OpenGothic/build/opengothic/libMoltenTempest.so(_ZN7Tempest11Application4execEv+0x1b) [0x7fd462b88faf]
#9: main - /home/black_fox/src/OpenGothic/build/opengothic/Gothic2Notr(main+0x121) [0x9dfd77]
```

First four stack frames are skipped to avoid printing the signal handling parts - the unskipped trace starts like this:
```
#1: CrashLog::tracebackLinux(std::ostream&) - /home/black_fox/src/OpenGothic/build/opengothic/Gothic2Notr(_ZN8CrashLog14tracebackLinuxERSo+0x36) [0xa47d56]
#2: CrashLog::dumpStack(char const*) - /home/black_fox/src/OpenGothic/build/opengothic/Gothic2Notr(_ZN8CrashLog9dumpStackEPKc+0xb2) [0xa47a30]
#3: /home/black_fox/src/OpenGothic/build/opengothic/Gothic2Notr() [0xa475fc]
#4: /lib64/libc.so.6(+0x3da60) [0x7f89b0a2aa60]
#5: InventoryMenu::keyUpEvent(Tempest::KeyEvent&) - /home/black_fox/src/OpenGothic/build/opengothic/Gothic2Notr(_ZN13InventoryMenu10keyUpEventERN7Tempest8KeyEventE+0x39) [0xa3b145]
(...)
```

File: Game/utils/crashlog.cpp [hunk 1]
```diff
@@ -92,25 +97,78 @@ void CrashLog::dumpStack(const char *sig) {
...
   fout << "---crashlog(" <<  sig << ")---" << std::endl;
   writeSysInfo(fout);
+#ifdef __WINDOWS__
   traceback.log(db, fout);
+#elif __LINUX__
+  tracebackLinux(fout);
 #endif
   }
 
+std::string CrashLog::demangle(const void* frame, const char* symbol) {
```
Comment (Try) [inline Game/utils/crashlog.cpp:126]: Using std::string in crash-log is risky: string do heap-allocation, while heap can be already corrupted. 
May be just emplace this code into `void CrashLog::tracebackLinux`

File: CMakeLists.txt [hunk 2]
```diff
@@ -103,6 +103,8 @@ target_link_libraries(${PROJECT_NAME} GothicShaders)
 
 if(NOT MSVC)
   target_compile_options(${PROJECT_NAME} PRIVATE -Wall -Wconversion -Wno-strict-aliasing -Werror)
+  # -rdynamic is needed to allow obtaining meaningful backtraces
+  target_link_options(${PROJECT_NAME} PRIVATE -rdynamic)
```
Comment (Try) [inline CMakeLists.txt:107]: `if(UNIX)` ?
-rdynamic seems not working in mingw

**Review by Try — CHANGES_REQUESTED** [review]

File: Game/utils/crashlog.cpp (hunk 1 above)
Comment (Nindaleth) [reply Game/utils/crashlog.cpp]: I agree, the reworked approach should prevent these allocations.

File: CMakeLists.txt (hunk 2 above)
Comment (Nindaleth) [reply CMakeLists.txt]: It's good there's a CI attached :-) Added UNIX check, should be much more specific.

**Comment by Nindaleth** [comment]
I've removed the [WIP] - the review comments have been hopefully resolved, please take another look.

**Comment by Try** [comment]
Looks good to me - merged.

**Pull request merged**